crates/noirc_evaluator/src/ssa/conditional.rs

```diff
@@ -577,6 +577,35 @@ impl DecisionTree {
                             DecisionTree::short_circuit(ctx, stack, ass_value, &error, *location)?;
                             return Ok(false);
                         }
+                    } else {
+                        let ass_value = Self::unwrap_predicate(ctx, &Some(ass_value));
+                        let op = Operation::Cast(ass_value);
+                        let pred = ctx.add_instruction(Instruction::new(
+                            op,
+                            ctx.object_type(*index),
+                            Some(stack.block),
+                        ));
+                        stack.push(pred);
+                        let op = Operation::Binary(node::Binary {
+                            lhs: *index,
+                            rhs: pred,
+                            operator: BinaryOp::Mul,
+                            predicate: None,
+                        });
+                        let idx = ctx.add_instruction(Instruction::new(
+                            op,
+                            ObjectType::native_field(),
+                            Some(stack.block),
+                        ));
+                        optimizations::simplify_id(ctx, idx).unwrap();
+                        stack.push(idx);
+
+                        let ins2 = ctx.instruction_mut(ins_id);
+                        ins2.operation = Operation::Load {
+                            array_id: *array_id,
+                            index: idx,
+                            location: *location,
+                        };
```

Comment:
So a conditional load `if c { a[i] }` is translated essentially to `a[c * i]` if I'm reading this correctly. Makes sense, but it is not obvious why we're doing it from the code alone so lets document it.

Reply:
I have added a comment, and also added a test to do it only when the index is not constant.